sdk/core/core-rest-pipeline/src/retryStrategies/throttlingRetryStrategy.ts

```diff
@@ -28,28 +34,93 @@ function parseRetryAfterHeader(headerValue: string): number | undefined {
   }
 }
 
+/**
+ * Returns the number of milliseconds to wait based on "retry-after-ms" or "x-ms-retry-after-ms" header value.
+ * Returns undefined if there is no valid value.
+ * @internal
+ */
+function parseRetryAfterMSHeader(retryAfterMSHeaderValue: string | undefined): number | undefined {
+  const delayValueString = retryAfterMSHeaderValue;
+
+  if (delayValueString == null) {
+    return;
+  }
+
+  const delayValueMs: number = Number(delayValueString);
+
+  if (Number.isNaN(delayValueMs)) {
+    return;
+  }
+
+  return delayValueMs;
+}
+
+/**
+ * Type denoting whether a reponse is throttling retry response along with the `retryAfterInMs` value if it is indeed a throttling retry response.
+ * @internal
+ */
+type ThrottlingInfoInternal =
+  | {
+      isThrottlingRetryResponse: false;
+    }
+  | { isThrottlingRetryResponse: true; retryAfterInMs: number };
+
+/**
+ * A response is a throttling retry response if it has a throttling status code (429 or 503),
+ * as long as one of the [ "Retry-After" or "retry-after-ms" or "x-ms-retry-after-ms" ] headers has a valid value.
+ *
+ * Returns an object denoting whether a reponse is throttling retry response along with the `retryAfterInMs` value if it is indeed a throttling retry response.
+ *
+ * @internal
+ */
+function getThrottlingInfo(response?: PipelineResponse): ThrottlingInfoInternal {
+  const defaultInfo: ThrottlingInfoInternal = {
+    isThrottlingRetryResponse: false,
+  };
+
+  if (!(response && [429, 503].includes(response.status))) return defaultInfo;
+
+  const retryAfterHeader = response.headers.get("Retry-After"); // in seconds
+
+  if (retryAfterHeader) {
+    const parsedRetryAfterHeader = parseRetryAfterHeader(retryAfterHeader);
+    if (parsedRetryAfterHeader) {
+      return { isThrottlingRetryResponse: true, retryAfterInMs: parsedRetryAfterHeader };
+    }
+  }
+
+  const retryAfterMSHeader =
+    response.headers.get(RetryAfterMillisecondsHeaders[0]) ?? // in milliseconds
+    response.headers.get(RetryAfterMillisecondsHeaders[1]); // in milliseconds
+
+  if (retryAfterMSHeader) {
+    const parsedRetryAfterHeader = parseRetryAfterMSHeader(retryAfterMSHeader);
+    if (parsedRetryAfterHeader) {
+      return { isThrottlingRetryResponse: true, retryAfterInMs: parsedRetryAfterHeader };
+    }
+  }
+
+  return defaultInfo;
+}
+
 /**
  * A response is a retry response if it has a throttling status code (429 or 503),
- * as long as the Retry-After header has a valid value.
+ * as long as one of the [ "Retry-After" or "retry-after-ms" or "x-ms-retry-after-ms" ] headers has a valid value.
  */
 export function isThrottlingRetryResponse(response?: PipelineResponse): boolean {
```

Comment:
nit: This function does a lot more calculation than necessary to determine if it is a throttling response, could it be optimized?

Reply:
Until it calculates the value of the `retry-after-in-ms` from the given response, we can't be sure if it is a throttling retry response.
Hence, we need to calculate the value before confirming.

Reply:
My point was that the math calculations are not strictly needed.

Is what is happening right now that `isThrottlingRetryResponse` gets called and then if it returns true, `throttlingRetryStrategy` gets called?

If so, then that seems a bit inefficient because `getThrottlingInfo` gets called twice with the same inputs. Is there a way to get rid of `isThrottlingRetryResponse` itself so the reasoning done in `getThrottlingInfo` is done once instead of twice?

Reply:
> Is what is happening right now that isThrottlingRetryResponse gets called and then if it returns true, throttlingRetryStrategy gets called?

No, the helper method is used somewhere else in exponential retry logic, hence retained the method.
It isn't used as part of throttlingRetryStrategy.

Reply:
Oh I see, thanks for clarifying!